File: modelsqlite/src/main/java/com/zpf/modelsqlite/SqlUtil.java
package com.zpf.modelsqlite;

import android.app.Application;
import android.database.sqlite.SQLiteOpenHelper;

import com.zpf.modelsqlite.constant.SQLiteConfig;
import com.zpf.modelsqlite.interfaces.ISqlJsonUtil;
import com.zpf.modelsqlite.interfaces.ObjCreator;
import com.zpf.modelsqlite.retrofit.SqlRetrofit;
import com.zpf.modelsqlite.utils.SqlCreatorImpl;
import com.zpf.modelsqlite.utils.SqlJsonUtilImpl;

import java.util.logging.Logger;

public class SqlUtil {
    private SqlUtil() {
    }

    private static volatile CacheDao defSqlDao;

    public static void initConfig(Application app, String dbPath) {
        SQLiteConfig.initConfig(app, dbPath);
    }

    public static CacheDao getDao() {
        if (defSqlDao == null) {
            synchronized (SqlUtil.class) {
                if (defSqlDao == null) {
                    defSqlDao = new CacheDao(SQLiteConfig.DB_PATH);
                }
            }
        }
        return defSqlDao;
    }

    public static CacheDao creatCacheDao(SQLiteOpenHelper openHelper) {
        return new CacheDao(openHelper);
    }

    public static void setJsonUtil(ISqlJsonUtil util) {
        SqlJsonUtilImpl.INSTANCE.setJsonUtil(util);
    }

    public static void addTypeCreator(ObjCreator<?> creator) {
        SqlCreatorImpl.INSTANCE.addTypeCreator(creator);
    }

    public static void setLogger(Logger logger) {
        com.zpf.modelsqlite.utils.Logger.INSTANCE.setRealLogger(logger);
    }

    public static <T> T create(Class<T> service) {
        return SqlRetrofit.INSTANCE.create(service);
    }
}

File: modelsqlite/src/main/java/com/zpf/modelsqlite/utils/SqlUtil.kt
package com.zpf.modelsqlite.utils

import android.app.Application
import android.content.pm.ApplicationInfo
import com.zpf.modelsqlite.*
import com.zpf.modelsqlite.constant.ColumnEnum
import com.zpf.modelsqlite.constant.SQLiteConfig
import java.lang.reflect.*
import java.lang.reflect.Array
import java.util.*
import kotlin.collections.HashMap

object SqlUtil {
    private val creatorMap = HashMap<Type, ObjCreator<*>>()

    internal val defJsonUtil: ISqlJsonUtil by lazy {
        SqlJsonUtilImpl()
    }

    private val defSqlDao: CacheDao by lazy {
        CacheDao(SQLiteConfig.DB_PATH)
    }

    fun initConfig(app: Application, tag: String? = null, dbPath: String? = null) {
        SQLiteConfig.DEBUG = ((app.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0)
        if (!tag.isNullOrEmpty()) {
            SQLiteConfig.LOG_TAG = tag
        }
        SQLiteConfig.appContext = app
        SQLiteConfig.DB_PATH = dbPath
    }

    fun get(): CacheDao {
        return defSqlDao
    }

    fun setJsonUtil(util: ISqlJsonUtil?) {
        (defJsonUtil as? SqlJsonUtilImpl)?.setJsonUtil(util)
    }

    fun addObjCreator(type: Type, creator: ObjCreator<*>) {
        creatorMap[type] = creator
    }

    fun newInstance(type: Type): Any? {
        val typeCreator = creatorMap[type]
        if (typeCreator != null) {
            return typeCreator.create(type)
        }
        val rawTypeClass = Utils.getRawTypeClass(type)
        var result: Any? = null
        if (rawTypeClass != null) {
            result = try {
                rawTypeClass.newInstance()
            } catch (e: Exception) {
                Logger.e(e.toString())
            }
            if (result == null) {
                try {
                    val constructor = rawTypeClass.getDeclaredConstructor()
                    val args: kotlin.Array<Any>? = null
                    constructor.newInstance(args)
                } catch (e: Exception) {
                    Logger.e(e.toString())
                }
            }
        }
        if (result != null) {
            return result
        }
        return result
    }


}